src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs

```diff
@@ -0,0 +1,424 @@
+                writeBioPtr = IntPtr.Zero,
+                isServer = isServer,
+            };
+
+            sslContext.isServer = isServer;
+
+            try
+            {
+                IntPtr method = GetSslMethod(isServer, options);
+
+                var contextPtr = libssl.SSL_CTX_new(method);
+                if (IntPtr.Zero == contextPtr)
+                {
+                    throw CreateSslException("Failed to allocate SSL/TLS context");
+                }
+
+                libssl.SSL_CTX_ctrl(contextPtr, libssl.SSL_CTRL_OPTIONS, options, IntPtr.Zero);
+
+                sslContext.sslPtr = libssl.SSL_new(contextPtr);
+
+                libssl.SSL_CTX_free(contextPtr);     
+
+                if (IntPtr.Zero == sslContext.sslPtr)
+                {
+                    throw CreateSslException("Failed to create SSSL object from SSL context");
+                }      
+
+                IntPtr memMethodRead = libcrypto.BIO_s_mem();
+                IntPtr memMethodWrite = libcrypto.BIO_s_mem();
+                if ((IntPtr.Zero == memMethodWrite) || (IntPtr.Zero == memMethodRead))
+                {
+                    throw CreateSslException("Failed to return memory BIO method function");
+                }
+
+                sslContext.readBioPtr = libssl.BIO_new(memMethodRead);
+                sslContext.writeBioPtr = libssl.BIO_new(memMethodWrite);
+                if ((IntPtr.Zero == sslContext.readBioPtr) || (IntPtr.Zero == sslContext.writeBioPtr))
+                {
+                    throw CreateSslException("Failed to retun new BIO for a given method type");
+                }
+
+                if (isServer)
+                {
+                    libssl.SSL_set_accept_state(sslContext.sslPtr);
+                }
+                else
+                {
+                    libssl.SSL_set_connect_state(sslContext.sslPtr);
+                }
+
+                libssl.SSL_set_bio(sslContext.sslPtr, sslContext.readBioPtr, sslContext.writeBioPtr);
+            }
+            catch
+            {
+                Marshal.StructureToPtr(sslContext, sslContextPtr, false);
+                FreeSslContext(sslContextPtr);
+                return IntPtr.Zero;
+            }
+            Marshal.StructureToPtr(sslContext, sslContextPtr, false);
+            return sslContextPtr;
+        }
+
+        internal static bool DoSslHandshake(IntPtr sslContextPtr, IntPtr recvPtr, int recvCount, out IntPtr sendPtr, out int sendCount)
+        {
+            sendPtr = IntPtr.Zero;
+            sendCount = 0;
+            SslContext context = Marshal.PtrToStructure<SslContext>(sslContextPtr);
+            bool isServer = context.isServer;
+            if ((IntPtr.Zero != recvPtr) && (recvCount > 0))
+            {
+                BioWrite(context.readBioPtr, recvPtr, recvCount);
+            }
+
+            int retVal = libssl.SSL_do_handshake(context.sslPtr);
+            if ((retVal == 1) && !isServer)
```

Comment:
Why `!isServer`?  Would a client auth cert connection need to do more work here?  Or is it really just a client side v server side distinction? Comments would help.